botblox_config/switch/switch.py

```diff
@@ -0,0 +1,688 @@
+
+    def _init_ports(self) -> None:
+        """
+        Initialize self._ports.
+        """
+        raise NotImplementedError()
+
+    def _init_registers(self) -> None:
+        """
+        Initialize self._registers.
+        """
+        raise NotImplementedError()
+
+    def _init_fields(self) -> None:
+        """
+        Initialize self._fields.
+        """
+        raise NotImplementedError()
+
+    def get_registers(self) -> Dict[int, Dict[int, 'Register']]:
+        """
+        :return: All registers of the switch.
+        """
+        return self._registers
+
+    def _add_register(self, register: 'Register') -> None:
+        """
+        Add the given register to this switch. This method should only be called during initialization.
+        :param register: The register to add.
+        """
+        if register.phy not in self._registers:
+            self._registers[register.phy] = dict()
+        self._registers[register.phy][register.mii] = register
+
+    def _add_field(self, field: 'MIIField') -> None:
+        """
+        Add the given field to this switch. This method should only be called during initialization.
+        :param field: The field to add.
+        """
+        self.fields[field.get_name()] = field
+
+    def _create_port_list_field(self, register: 'Register', index: int, ports_default: bool, name: str)\
+            -> 'PortListField':
+        """
+        Create a field that represents a bitmask (or other representation) of a set/list of ports.
+        :param register: The register that backs this list.
+        :param index: Index in the register (implementation-specific).
+        :param ports_default: Whether all ports should be set by default or not.
+        :param name: Name of the field.
+        :return: The field.
+        """
+        raise NotImplementedError()
+
+    def check_vlan_id(self, vlan_id: int) -> None:
+        """
+        Check whether the given VLAN ID is valid.
+        :param vlan_id: The ID to check.
+        :raise ValueError: If it's invalid.
+        """
+        if not 0 <= vlan_id < self.max_vlan_id():
+            raise ValueError("VLAN ID {} is invalid for {} with max VLAN ID {}".format(
+                             vlan_id, self.name(), self.max_vlan_id()))
+
+    def has_feature(self, feature: SwitchFeature) -> bool:
+        """
+        Check if this switch has the given feature.
+        :param feature: The feature to check.
+        :return: Whether the switch has the feature.
+        """
+        return feature in self._features
+
+    def check_feature(self, feature: SwitchFeature) -> None:
+        """
+        Check that this switch has the given feature.
+        :param feature: The feature to check.
+        :raise RuntimeError: If the switch does not have the feature.
+        """
+        if not self.has_feature(feature):
+            raise RuntimeError("{} does not support {}".format(self.name(), feature.value))
+        pass
+
+
+class Register:
```

Comment:
This is a good class as it clearly abstracts a lot of the common operations that are being performed on the register data. The challenge is two-fold: (1) Currently the setup is that this register is very unique to the SMI protocol (as it should be as that's the protocol we use in firmware to communicate with the ip175g) but what happens if this expands to SPI, I2C as it could in future? You could argue for a separate Register class (RegisterSPI, etc) or you could refactor Register to handle all cases. 

(2) Is that I'm a little unsure about the idea of using the `_init_registers` method in switch chip to add the registers in a hardcoded way - I suppose one of the advantages of the issue made about putting information in a `.json` file is that all the information that you need to configure the IP175G is in one place and self-documented in a similar style to the dictionary of register values used in the `port mirroring` and `vlan` cli configs. Would you be open to the idea of refactoring along those lines? 

Another comment is should we be making classes to represent individual bits, bytes and 2-byte data values? Maybe? The pros are that it lends itself to this `fields` attribute, which corresponds really well with what you would read in the datasheet. The downside is that it seems a bit arbitrary - alternatively I could acheive the same result with all of them using a bitmask and an offset on the whole two-byte data. Also I wonder if they pay for themselves in terms of reducing complexity of the code?
Tell you the truth, I'm not sure either way (one to discuss). 

I hope I am making sense? Please correct me if I am mistaken

Reply:
I think that calling it `SMIRegister` or `MIIRegister` might be good for now. We can refactor for other protocols later SPI and I2C later and when needed.

Reply:
I kept the `Register` class as a base and added `MIIRegister` as the particular implementation.